### DISCUSSION TITLE
Problems with the "build_opt.h" setting

### DISCUSSION THREAD

**Discussion by JingtaiChen** [stm32duino/Arduino_Core_STM32#2297] [Q&A]
I saw the wiki's Core Callback about the API, which required me to create a build opt in the same folder as the ino file. I followed h's file. As the picture shows:
<img width="939" alt="image" src="https://github.com/stm32duino/Arduino_Core_STM32/assets/111559992/9cd9c8b7-ed6c-4bb7-ade8-a17e6cfc7143">
<img width="273" alt="image" src="https://github.com/stm32duino/Arduino_Core_STM32/assets/111559992/ed4d3595-0366-4659-98e1-d52eebc5b7d5">
But I have a problem when I call the HAL library to interrupt the serial module：
`HAL_StatusTypeDef HAL_UART_Receive_IT(UART_HandleTypeDef *huart, uint8_t *pData, uint16_t Size)`
Wiki tells me to write:
`void registerCoreCallback(void (*func)(void))`
How do I write the function "HAL_UART_Receive_IT" in the HAL module as "void registerCoreCallback(void (*func)(void)") as required in the Wikipedia

**Comment by JingtaiChen** [comment]
I tried to modify the ino file
`UART_HandleTypeDef huart1;
uint8_t Message[1];

void registerCoreCallback(void HAL_UART_RxCpltCallback(UART_HandleTypeDef *huart));

void HAL_UART_RxCpltCallback(UART_HandleTypeDef *huart) {
  if (huart == &huart1)
  {
    HAL_UART_Transmit_IT(&huart1, Message, 1);
    HAL_UART_Receive_IT(&huart1, Message, 1);
  }
}
void setup() {
  __HAL_RCC_GPIOA_CLK_ENABLE();
  huart1.Instance = USART1;
  huart1.Init.BaudRate = 115200;
  huart1.Init.WordLength = UART_WORDLENGTH_8B;
  huart1.Init.StopBits = UART_STOPBITS_1;
  huart1.Init.Parity = UART_PARITY_NONE;
  huart1.Init.Mode = UART_MODE_TX_RX;
  huart1.Init.HwFlowCtl = UART_HWCONTROL_NONE;
  huart1.Init.OverSampling = UART_OVERSAMPLING_16;
  if (HAL_UART_Init(&huart1) != HAL_OK) {
  }
  HAL_UART_Receive_IT(&huart1, Message, 1);
}

void loop() {
}

`
Then my build_opt.h file looks like this:
`
-DHAL_UART_MODULE_ENABLED
-DCORE_CALLBACK
`
But he reported the following malfunction
<img width="1787" alt="image" src="https://github.com/stm32duino/Arduino_Core_STM32/assets/111559992/6c876701-0d32-4205-b102-56bc44f39504">
How can I fix it？

**Reply by fpistm** [reply]
You can't redefine those handler. Core callback is for custom cb not the one for HAL.

**Comment by JingtaiChen** [comment]
So how do I call something like HAL_StatusTypeDef HAL_UART_Receive_IT(UART_HandleTypeDef *huart, uint8_t *pData, uint16_t Size), In response to a serial port interruption?

**Reply by fpistm** [reply]
This function is part of the HAL. You can try to call it yourself but don't advise nor understand why you want to use it. It is used by the HardwareSerial implementation.

**Reply by fpistm** [reply]
If you want to do your own U(S)ART implementation simply define `HAL_UART_MODULE_ONLY` in build_opt.h like this:
`-DHAL_UART_MODULE_ONLY`

https://github.com/stm32duino/Arduino_Core_STM32/wiki/HAL-configuration#hal-module-only

**Reply by JingtaiChen** [reply]
Because I want to use the DMA response interrupt of the serial port, I hope to simply use the configuration of STM32CubeMX to form the configuration code, and then put it in ArduinoIDE. When the serial port needs high-speed serial transmission, I use the syntax of CubeIDE to call the DMA module. When high-speed serial transmission is not required, the original Arduino API is used. In this case, I just use the serial port interrupt as an experiment: it is to use the Cubeide-like syntax in ArduinoIDE environment to call the serial port callback function.
In addition, I changed the build_opt.h to the following：
<img width="323" alt="image" src="https://github.com/stm32duino/Arduino_Core_STM32/assets/111559992/24843a70-c6c7-4870-937c-078b75fc5bb3">
But he still reports the following faults：
<img width="964" alt="image" src="https://github.com/stm32duino/Arduino_Core_STM32/assets/111559992/070255dd-4e00-48f0-b082-3e1be5feffb2">

**Reply by fpistm** [reply]
The error is clear. Multiple definition. 
Just to be clear core callback has no ling with hal register callback feature.

**Reply by JingtaiChen** [reply]
I see, but since ArduinoIDE automatically calls the STM32Hal definitions, how do I make the "HAL_UART_Receive_IT" function only be called from my code and not from Arduino?

**Reply by fpistm** [reply]
https://github.com/orgs/stm32duino/discussions/2297#discussioncomment-8703647

**Reply by JingtaiChen** [reply]
I changed the "build_opt.h" file to the following according to your description:
<img width="436" alt="image" src="https://github.com/stm32duino/Arduino_Core_STM32/assets/111559992/eebc76a5-d2da-4f77-8163-1c40084517de">
And change the program to the following:
`UART_HandleTypeDef huart1;
uint8_t Message[1];

void setup() {

  __HAL_RCC_GPIOA_CLK_ENABLE();
  huart1.Instance = USART1;
  huart1.Init.BaudRate = 115200;
  huart1.Init.WordLength = UART_WORDLENGTH_8B;
  huart1.Init.StopBits = UART_STOPBITS_1;
  huart1.Init.Parity = UART_PARITY_NONE;
  huart1.Init.Mode = UART_MODE_TX_RX;
  huart1.Init.HwFlowCtl = UART_HWCONTROL_NONE;
  huart1.Init.OverSampling = UART_OVERSAMPLING_16;
  HAL_UART_Init(&huart1);

  // HAL_UART_Receive_IT(&huart1, Message, 1);
}

void loop() {
  uint8_t Message_A[] = "Hello CubeIDE\n";
  HAL_UART_Transmit(&huart1, (uint8_t *)Message_A, 14, 100);

  uint8_t Message_B[] = "Hello CubeIIT\n";
  HAL_UART_Transmit_IT(&huart1, (uint8_t *)Message_B, 14);

}

// void HAL_UART_RxCpltCallback(UART_HandleTypeDef *huart) {
//   if (huart == &huart1) {
//     HAL_UART_Transmit_IT(&huart1, (uint8_t *)Message, 1);
//     HAL_UART_Receive_IT(&huart1, Message, 1);
//   }
// }`
But the program is no longer working, but if you change the program to:
`UART_HandleTypeDef huart1;
uint8_t Message[1];

void setup() {

  __HAL_RCC_GPIOA_CLK_ENABLE();
  huart1.Instance = USART1;
  huart1.Init.BaudRate = 115200;
  huart1.Init.WordLength = UART_WORDLENGTH_8B;
  huart1.Init.StopBits = UART_STOPBITS_1;
  huart1.Init.Parity = UART_PARITY_NONE;
  huart1.Init.Mode = UART_MODE_TX_RX;
  huart1.Init.HwFlowCtl = UART_HWCONTROL_NONE;
  huart1.Init.OverSampling = UART_OVERSAMPLING_16;
  HAL_UART_Init(&huart1);

  // HAL_UART_Receive_IT(&huart1, Message, 1);
}

void loop() {
  uint8_t Message_A[] = "Hello CubeIDE\n";
  HAL_UART_Transmit(&huart1, (uint8_t *)Message_A, 14, 100);

  // uint8_t Message_B[] = "Hello CubeIIT\n";
  // HAL_UART_Transmit_IT(&huart1, (uint8_t *)Message_B, 14);

}

// void HAL_UART_RxCpltCallback(UART_HandleTypeDef *huart) {
//   if (huart == &huart1) {
//     HAL_UART_Transmit_IT(&huart1, (uint8_t *)Message, 1);
//     HAL_UART_Receive_IT(&huart1, Message, 1);
//   }
// }`
The program output: Hello CubeIDE,This indicates that HAL_UART_Transmit_IT still cannot be called. Why is that?
